code/controllers/subsystems/mapping.dm

```diff
@@ -44,37 +57,192 @@ SUBSYSTEM_DEF(mapping)
+
+	// Now lets start choosing some.
+	while(budget > 0 && overall_sanity > 0)
+		overall_sanity--
+		var/datum/map_template/chosen_template = null
+
+		if(potential_submaps.len)
+			if(priority_submaps.len) // Do these first.
+				chosen_template = pick(priority_submaps)
+			else
+				chosen_template = pick(potential_submaps)
+
+		else // We're out of submaps.
+			log_mapload("Submap loader had no submaps to pick from with [budget] left to spend.")
+			break
+
+		CHECK_TICK
+
+		// Can we afford it?
+		if(chosen_template.cost > budget)
+			priority_submaps -= chosen_template
+			potential_submaps -= chosen_template
+			continue
+
+		// Did we already place down a very similar submap?
+		if(chosen_template.template_group && (chosen_template.template_group in template_groups_used))
+			priority_submaps -= chosen_template
+			potential_submaps -= chosen_template
+			continue
+
+		// If so, try to place it.
+		var/time_started_submap = REALTIMEOFDAY
+		var/specific_sanity = 100 // A hundred chances to place the chosen submap.
+		while(specific_sanity > 0)
+			specific_sanity--
+
+			var/orientation
+			if(chosen_template.fixed_orientation || !config.random_submap_orientation)
+				orientation = 0
+			else
+				orientation = pick(list(0, 90, 180, 270))
+
+			chosen_template.preload_size(chosen_template.mappath, orientation)
+			var/width_border = TRANSITIONEDGE + SUBMAP_MAP_EDGE_PAD + round(((orientation%180) ? chosen_template.height : chosen_template.width) / 2) // %180 catches East/West (90,270) rotations on true, North/South (0,180) rotations on false
+			var/height_border = TRANSITIONEDGE + SUBMAP_MAP_EDGE_PAD + round(((orientation%180) ? chosen_template.width : chosen_template.height) / 2)
+			var/z_level = pick(z_levels)
+			var/turf/T = locate(rand(width_border, world.maxx - width_border), rand(height_border, world.maxy - height_border), z_level)
+			var/valid = TRUE
+
+			for(var/turf/check in chosen_template.get_affected_turfs(T,TRUE,orientation))
+				var/area/new_area = get_area(check)
+				if(!(istype(new_area, whitelist)))
+					valid = FALSE // Probably overlapping something important.
+			//		to_world("Invalid due to overlapping with area [new_area.type] at ([check.x], [check.y], [check.z]), when attempting to place at ([T.x], [T.y], [T.z]).")
+					break
+				CHECK_TICK
+
+			CHECK_TICK
+
+			if(!valid)
+				continue
+
+			admin_notice("Submap \"[chosen_template.name]\" placed at ([T.x], [T.y], [T.z])\n", R_DEBUG)
+
+			// Do loading here.
+			chosen_template.load(T, centered = TRUE, orientation=orientation) // This is run before the main map's initialization routine, so that can initilize our submaps for us instead.
+
+			var/time_ended_submap = (REALTIMEOFDAY - time_started_submap) / 10
+			if(verbose_loading_time_logs)
+				log_mapload("Loaded submap '[chosen_template.name]' in [time_ended_submap] second\s.")
+
+			CHECK_TICK
+
+			// For pretty maploading statistics.
+			if(loaded_submap_names[chosen_template.name])

```

Comment:
`loaded_submap_names[chosen_template.name] = loaded_submap_names[chosen_template.name] + 1` will do this as well.